### PR TITLE
Supermatter port

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports over TG's supermatter. 

So begins the ending of the age of the Singulo and Tesla

TG's supermatter fully replaces the singulo and Tesla. You will still be able to construct the singulo and Tesla, but it being a round-start engine available for players is gone (order it from cargo). 

The supermatter is an engine that interacts with gas, pressure, and temperature, while giving off radiation. For those of you interested in the finer details of the engine, I recommend TG's guide: https://tgstation13.org/wiki/Guide_to_the_Supermatter

The engine can be ran "hot" to intentionally produce anomalies, which can then be deactivated and given to science for bonuses in research. Of course, the anomalies themselves are dangerous, and if your'e not careful, things could spiral out of control and the engine could ultimately get away from you in your attempt to engage in this.

If the supermatter reaches critical point, it'll delaminate, causing a hug explosions while bathing the station in radiation and causing everyone to hallucinate.

If you managed to have a lot of gas in the chamber of the supermatter when it delaminates, it can compress down into a super singularity, which will begin consuming the station.

Alternatively, if the supermatter is particularly engergetic when it delamintes, it'll explode into an energy ball that'll begin shocking the station into oblivion.

pic:
img

:cl: Fox McCloud
add: Updates the supermatter
/:cl:

### CODE REVIEW

File: code/game/objects/effects/anomalies.dm [hunk 1]
```diff
@@ -5,43 +5,74 @@
...
+
+/obj/effect/anomaly/process()
+	anomalyEffect()
+	if(death_time < world.time)
+		if(loc)
+			detonate()
+		qdel(src)
 
 /obj/effect/anomaly/Destroy()
-	QDEL_NULL(aSignal)
```
Comment (Fox-McCloud) [inline code/game/objects/effects/anomalies.dm]: re-add this.

**Comment by KasparoVy** [comment]
SSIIIICK MANNNN

Time to learn how to SM. I'm going to miss the singulo (not really the tesla) but SM is gonna be interesting. Ready to see some next level shit

**Comment by Mitchs98** [comment]
Kinda happy for this. SM isn't nearly as round ending if it explodes. Sure it's a disaster but everything it causes can be repaired or prevent much easier than a singularity/tesla eating everything.

**Comment by Fox-McCloud** [comment]
>Kinda happy for this. SM isn't nearly as round ending if it explodes

Oh how wrong you are. TG's implementation of the supermatter doesn't necessarily just explode. It can explode AND result in a supermatter powered singularity or a tesla balls, depending on the gas composition and energy rating of the supermatter when it blows.

If a traitor knows what he's doing or engineering is especially dumb, it's arguably more destructive as it'll blow up, irradiate everyone, make them hallucinate, and potentially spawn a supermatter singularity or tesla energy ball.

**Comment by KasparoVy** [comment]
A lot more depth, more creative ways to totally waste the place!

**Comment by Fox-McCloud** [comment]
Anomalies and countdown tested; now to actually test the supermatter itself.

**Comment by Mitchs98** [comment]
> > Kinda happy for this. SM isn't nearly as round ending if it explodes
> 
> Oh how wrong you are. TG's implementation of the supermatter doesn't necessarily just explode. It can explode AND result in a supermatter powered singularity or a tesla balls, depending on the gas composition and energy rating of the supermatter when it blows.
> 
> If a traitor knows what he's doing or engineering is especially dumb, it's arguably more destructive as it'll blow up, irradiate everyone, make them hallucinate, and potentially spawn a supermatter singularity or tesla energy ball.

Oh. Well. Dang. lol. Oh well, at-least it'll take far more effort to do something malicious with it?

**Comment by farie82** [comment]
Only issue I can think of instantly is the amount of GOATS or early shuttle calls that will occur once this is in. Especially if the wiki page is not yet fully up to date on the day of the merge

**Comment by Citinited** [comment]
HOW DARE YOU 🔪

Just kidding haha, bring this on! The supermatter (even in paradise's current gimped version) is many times more interesting than either the singularity or the tesla.

I'm going to miss the singularity when this gets merged, though - will the required machinery be left in secure storage, or will it just be a cargo-only thing? I guess I can just filch the one from the construction site 😄

**Comment by Ralta** [comment]
Hoo-boy, this is gonna be interesting. I love the Supermatter, but there are gonna be a _lot_ of meltdowns initially. 

I wonder if it's worth _temporarily_ adding a couple of super-cooled gas canisters to secure storage at roundstart to give engineers/atmos techs a method of preventing a meltdown ~~if~~ when they mess up.

I appreciate the argument of `HUGBOX REEEEE` but the SM is very different to standard engines, and has a very high risk associated with it. Giving people time to acclimate and potentially fix issues rather than ending rounds prematurely seems like a good idea. 

Also assume there's gonna be a good amount of auditing on SM interaction to make it easy for admins to track griefing?

**Comment by Citinited** [comment]
Personally I don't feel that supercold gas canisters for the purpose of preventing meltdowns need to be added. Firstly, the solars exist for this reason - if people don't feel comfortable setting up the engine then they can wire the solars instead. And secondly, this will give something for atmospherics to do, should it be desired for supercooled gas to be made available.

**Comment by Ralta** [comment]
> Personally I don't feel that supercold gas canisters for the purpose of preventing meltdowns need to be added. Firstly, the solars exist for this reason - if people don't feel comfortable setting up the engine then they can wire the solars instead. And secondly, this will give something for atmospherics to do, should it be desired for supercooled gas to be made available.

It was suggested as a temporary measure while people get to grips with the SM; having them at roundstart means that when it inevitably starts delaminating, they can immediately halt the delam and find the problem with plenty of time to spare. If they fuck it up hard enough, the SM can delam super fast, and the outcome is usually a singularity. 

Imagine releasing the engine every other round. That's what it will be, I guarantee it. Whether or not anything should be done mechanically to prevent that up for debate I suppose, but it's a consideration.

**Comment by Citinited** [comment]
It is a valid concern; while my opinion on it is unchanged it probably does no harm to add them as an interim measure.

If we see the engine delaminate literally every round I'll probably revise my opinion on it 😄

**Comment by Ralta** [comment]
> It is a valid concern; while my opinion on it is unchanged it probably does no harm to add them as an interim measure.
> 
> If we see the engine delaminate literally every round I'll probably revise my opinion on it

My opinion on this is coming from experience with Oracle having the SM as the primary engine; a lot of our players came from Para, and I shit you not the SM would delaminate every other round initially. After a few weeks, it got much less frequent and it became much more obvious when someone was griefing instead of just incompetent. But Oracle also had access to Freon for supercooling in emergency situations. It's where I'm stealing the idea from :eyes:

Not saying it should be long term by any means; just something to stave off station destruction for a little bit while people get used to the construction. 

If nothing else, I'm looking forward to seeing more cooperation between atmos and engineering with creating supercooled canisters for the engine. I wonder if Atmos Techs should get engine access, in that case? :thinking:

**Comment by Citinited** [comment]
Yeah I think engine access would be rather useful for atmospherics techs - in addition to repiping the engine, they can do a bunch of other things such as setting up the cooling, filtering gases off, cooling the plasma inside the radiation collectors, things like that.

Does this version of the supermatter rely on a TEG for power output or does it go off radiation collectors?

**Comment by Fox-McCloud** [comment]
>Does this version of the supermatter rely on a TEG for power output or does it go off radiation collectors?

Rad collectors.

Using the TEG for power off of the SM is really weird and kinda negates the whole purpose of running the SM in particular ways to get particular effects out of it.

**Comment by taukausanake** [comment]
I don't know, I think I liked TEG for supermatter and nothing else really. I did also like SM using rad collectors. You still have to filter out hot gas while providing it cold gas don't you? Maybe running rad collectors and a TEG loop would be for a slightly more advanced engineering setup

**Comment by davidchan13** [comment]
@taukausanake You still can hook up the TEG to /tg/ SM, though at that point you already have all the power you'd ever need. Turbine is far more efficient for TEG loops anyways since you can get the gas much hotter without the danger of blowing all of engineering.

Overall, fairly neutral on this PR. Having round start SM would be very entertaining, I foresee a lot of rounds ending early due to incompetence or attempting dangerous builds without proper setup. I regularly see other tgcode stations going critical and early evac despite having had the engine for months, higher populations and the astonishing lack of atmos experience across paradise makes me hesitant to urge this forward.

@AffectedArc07 Perhaps your power transfer beacon/laser could have some use after all, moving the engineering outpost to the mining asteroid (and taking away their antag safe space) and having the SM on the roid with engineering shuttling down gas canisters for the SM, at least until more people get the hang of the 'new' engine and we don't risk 1/4+ of all rounds ending in early disaster before most antags can even plan how to tackle their objectives.

**Comment by KasparoVy** [comment]
I'm down for absolute engine mayhem.

**Comment by Mitchs98** [comment]
> moving the engineering outpost to the mining asteroid (and taking away their antag safe space)

The engineering outpost is where it is as it's one of if not the only reliable way to access the construction site, which people frequently use to teach newbie engineering/do fun projects on/cultify. To move it/remove it would require remapping where the construction site is at, which I see pointless to do.

**Comment by Fox-McCloud** [comment]
I got a bit overly ambitious and went ahead and mapped this in.

**Comment by KasparoVy** [comment]
Beast! Think we can get a screenshot or two of the thing in action?

**Comment by Fox-McCloud** [comment]
Yeah, just need to finish out testing; will probably get screenshots up tomorrow evening.

**Comment by Dinarzad** [comment]
So I may have missed this in the discussion but may as well ask:
Does this remove the singularity and tesla generators altogether or just from their pre-assembled round-start state?

I ask mostly because it'd be a shame to see Singuloth vanish into the ages if that's the case (Though I suppose it can always be ordered from cargo), though I am overall in favor of transitioning to the SM engine. The initial weeks of doing so are gonna lead to a lot of kabooms but in the long run once people adapt it should lead to something more interesting for engineers to work on.

**Comment by taukausanake** [comment]
Following @Dinarzad's comment, I think you are forgetting an image of the map. To run SM you normally need a closed area, and last I remember the singlo/tesla containment was in space. This also means that if the map is being designed for SM, then tesla and singlo (if they are meant to be used anymore after this) will also have to be in the closed area. Makes me wonder if either of the old engines would be safe anymore depending on how Engineering is redesigned

**Comment by SteelSlayer** [comment]
I like the idea of having a dedicated supermatter engine setup at round start. However I don't like the idea of completely removing an area for the tesla and singularity, if that is your intent. I feel it should be more like delta where there's an area for singularity/tesla and an area for the SM. Then engineering will have a choice between all three and can even do things like run the singularity/tesla AND the SM at the same time.

**Comment by Fox-McCloud** [comment]
updated with pic.

**Comment by Citinited** [comment]
Hmm. I have some feedback / questions on the layout of the engine in that screenshot:
1. Can the pipes between the scrubbers/vents and the pumps be visible? They lead directly into a wall and it's hard to tell where they go without getting out a t-ray scanner. (My preference would be to not have them go through walls in the first place but that's just me)
2. It seems weird that you'd use only a single pane of plasma glass for the core instead of the double-pane ones you see everywhere else. I see why you did it that way, but it looks.. odd. 
3. How many radsuits are mapped in for this? I can only see two in there.
4. The UVA in the bottom left of the engine room - does that lead into station filter loop?
5. No cycling airlock on the core? (e.g. like the ones at the turbine)

**Comment by Fox-McCloud** [comment]
>1. Can the pipes between the scrubbers/vents and the pumps be visible?

They are in terms of mapping capacity. Walls says "bite me", however.

>2. It seems weird that you'd use only a single pane of plasma glass for the core instead of the double-pane ones you see everywhere else. I see why you did it that way, but it looks.. odd. 

Price we pay for keeping everything compact and neat while allowing for expandability of future rad collectors

>3. How many radsuits are mapped in for this? I can only see two in there.

There's multiple suits to a closet. Also a few on the table

>4. The UVA in the bottom left of the engine room - does that lead into station filter loop?

Universal Pipe Adapter you mean? Yes.

>5. No cycling airlock on the core? (e.g. like the ones at the turbine)

pls nu make me work with the cancer that is airlock controllers.

**Comment by Citinited** [comment]
> pls nu make me work with the cancer that is airlock controllers.

I feel your pain 😄

**Comment by taukausanake** [comment]
If only we had that automatic airlock blueprint thing that made them more you :eyes:

**Comment by Citinited** [comment]
Indeed, I'm just waiting for Ansari to reopen it, but I think he's been busy recently. 👀

**Comment by Muhdi345** [comment]
It's about time! I will fiddle this fondly. Hope we'll get the exotic TG gasses too(in my dreams)...
And there's enough space to plop a TEG in just for giggles.

**Comment by realtaac** [comment]
i thought sm wasn't used because our atmos system didn't like it.

**Comment by taukausanake** [comment]
I think Fastmos kind of make it workable but I'm not too sure what kind of bastardization of atmos code is running right now. Sure does make me miss the glory days of ZAS though!

**Comment by Fox-McCloud** [comment]
>glory days

>>ZAS

Man, some people are getting pretty good at telling jokes these days!

**Comment by KasparoVy** [comment]
Man I think I got to experience ZAS all of once and I only realized what it was because some fuckin guy cracked a window to catch a breeze and zoopty woop, my ass was in space before I was even in the door

I also think it was on the prison server but my memory sucks

**Comment by Fox-McCloud** [comment]
>What a diZASter

I'm keeping this.

**Comment by taukausanake** [comment]
THAT'S what I loved about ZAS! You have any kind of breach and you get thrown from the window to the fucking wall. It gave _actual_ danger to breaches into space. I guess it wasn't all amazing that gases only spread in their own area and immediately at that, but damn did I have fun dying when it came to depressurization

**Comment by davidchan13** [comment]
Back when every exterior window had to have a fire shutter because oh my honkmother Carp migrations were dangerous.

**Comment by Citinited** [comment]
Oh god I remember every round when arrivals would have a random 1 * 1 breach and the big yellow firelocks would come down and trap everyone in arrivals behind opaque airlocks. Good times 😄

**Comment by Taaketa** [comment]
@Fox-McCloud  Looks like the emitters room is missing cameras?

**Comment by MrMagolor** [comment]
What about the particle accelerator for even setting up your own singulo/tesla? Is that in secure storage?

Also, I feel like that changelog is not accurate at all, you could at least mention that it replaces singulo/tesla on the cyberiad.

**Comment by Markolie** [comment]
Would really love for this to make it in.

**Comment by Fox-McCloud** [comment]
>What about the particle accelerator for even setting up your own singulo/tesla? Is that in secure storage?

You can still order the parts for the Tesla and Singulo via cargo, but they are not in storage, by default.

**Comment by Kyep** [comment]
Okay, so I spent a few hours playing around with this today.

I still think it is necessary to retain support for roundstart tesla/singulo engines, because:

1. Many people can't even set up the tesla/singulo correctly. Asking them to set up the much more complicated SM is... going to result in a LOT of shifts ruined by supermatter delamination.
2. It gives people time to learn the new engine at their own pace.
3. It allows for potential tradeoffs to be made between the engines in future PRs.

That aside, from testing it and looking at the map, here is my feedback:

engines

1. The grid power monitor on the left side (left red box) is no longer connected to the engine power grid, and thus does not work. It needs to be moved to somewhere that is (e.g: red arrow)
2. There is no emitter-specific SMES (green box). This is a problem because it means that if there is no power in the grid the emitters won't fire and the engine won't start. I suggest putting in a SMES both so that the engine can be started if the grid has no power, and also so that power to the emitters (cyan box) can be easily cut in an emergency without cutting power to the whole grid.
3. There is a pacman generator at the top right (red box next to SMES) but it has no wire under it and thus doesn't work by default.
4. It might be a good idea to give eng another gas canister or two, for various purposes (emergency cooling, sabotage, etc).

**Comment by TDSSS** [comment]
can you cheese the SM with a atmos holobarrier on its tile, to prevent any nasty gases from building up?

**Comment by AffectedArc07** [comment]
Holofanning the supermatter will stop the plasma from escaping from it, and also stop the nitrogen hitting it. Holofanning the SM will **make it blow up**

**Comment by Markolie** [comment]
The supermatter damage overlay will need to have its mouse_opacity set to fully transparent so you can examine the supermatter. Just a small thing that came up today.

File: code/modules/power/supermatter/supermatter.dm [hunk 2]
```diff
@@ -1,395 +1,588 @@
...
+			if(damage > emergency_point)
+				radio.autosay("[emergency_alert] Integrity: [get_integrity()]%", name, null, list(z))
+				lastwarning = REALTIMEOFDAY
+				if(!has_reached_emergency)
+					investigate_log("has reached the emergency point for the first time.", "supermatter")
+					message_admins("[src] has reached the emergency point [ADMIN_JMP(src)].")
+					has_reached_emergency = TRUE
+			else if(damage >= damage_archived) // The damage is still going up
+				radio.autosay("[warning_alert] Integrity: [get_integrity()]%", name, "Engineering", list(z))
+				lastwarning = REALTIMEOFDAY - (WARNING_DELAY * 5)
```
Comment (farie82) [inline code/modules/power/supermatter/supermatter.dm:454]: Why the `* 5` here? Just a minus will also make it warn engineering every processing tick

File: code/modules/power/supermatter/supermatter.dm [hunk 3]
```diff
@@ -1,395 +1,588 @@
...
+				if(powerloss_inhibitor < 0.5)
+					radio.autosay("DANGER: CHARGE INERTIA CHAIN REACTION IN PROGRESS.", name, "Engineering", list(z))
 
+			if(combined_gas > MOLE_PENALTY_THRESHOLD)
+				radio.autosay("Warning: Critical coolant mass reached.", name, "Engineering", list(z))
 
+		if(damage > explosion_point)
+			countdown()
+
+	return 1
```
Comment (farie82) [inline code/modules/power/supermatter/supermatter.dm:471]: TRUE

File: code/modules/power/supermatter/supermatter.dm [hunk 4]
```diff
@@ -1,395 +1,588 @@
...
-/obj/machinery/power/supermatter_shard/attackby(obj/item/W as obj, mob/living/user as mob, params)
-	if(!istype(W) || (W.flags & ABSTRACT) || !istype(user))
+/obj/machinery/power/supermatter_crystal/attackby(obj/item/I, mob/living/user, params)
+	if(!istype(I) || (I.flags & ABSTRACT) || !istype(user))
 		return
-	if(user.drop_item(W))
-		Consume(W)
-		user.visible_message("<span class='danger'>As [user] touches \the [src] with \a [W], silence fills the room...</span>",\
-			"<span class='userdanger'>You touch \the [src] with \the [W], and everything suddenly goes silent.\"</span>\n<span class='notice'>\The [W] flashes into dust as you flinch away from \the [src].</span>",\
+	if(moveable && default_unfasten_wrench(user, I, time = 20))
```
Comment (farie82) [inline code/modules/power/supermatter/supermatter.dm:567]: Needs to be moved to the wrench_act proc

File: code/modules/power/supermatter/supermatter.dm [hunk 5]
```diff
@@ -398,77 +591,173 @@
...
+		user.dust()
+		matter_power += 200
+	else if(istype(AM, /obj/singularity))
+		return
+	else if(istype(AM, /obj))
+		if(!istype(AM, /obj/effect))
+			var/suspicion = ""
+			if(AM.fingerprintslast)
+				suspicion = "last touched by [AM.fingerprintslast]"
+				message_admins("[src] has consumed [AM], [suspicion] [ADMIN_JMP(src)].")
```
Comment (farie82) [inline code/modules/power/supermatter/supermatter.dm:610]: won't this spam admins?

File: code/modules/power/supermatter/supermatter.dm [hunk 6]
```diff
@@ -398,77 +591,173 @@
...
-	power += 200
-	supermatter_zap()
-
+	if(!istype(AM, /obj/effect))
+		matter_power += 200
 
 	//Some poor sod got eaten, go ahead and irradiate people nearby.
 	for(var/mob/living/L in range(10))
-		var/rads = 500 * sqrt( 1 / (get_dist(L, src) + 1) )
+		var/rads = 500 * sqrt( 1 / (get_dist(L, src) + 1))
```
Comment (farie82) [inline code/modules/power/supermatter/supermatter.dm:618]: Can cause a divided by zero runtime if the mob in question is on the same tile as the SM. Might be possible if you open the crate when the mob is on/under the crate

File: code/game/objects/effects/anomalies.dm [hunk 7]
```diff
@@ -60,72 +91,188 @@
...
 		boing = 0
-		return
 
 /////////////////////
 
 /obj/effect/anomaly/flux
 	name = "flux wave anomaly"
 	icon_state = "electricity2"
+	density = TRUE
+	var/canshock = 0
```
Comment (farie82) [inline code/game/objects/effects/anomalies.dm:131]: Boolean value so best to use FALSE/TRUE

File: code/game/objects/effects/anomalies.dm [hunk 8]
```diff
@@ -60,72 +91,188 @@
...
+				if(istype(A, /obj/item/radio/beacon))
+					continue // don't teleport beacons because that's just insanely stupid
+				if(A.anchored)
+					continue
+
+				var/turf/newloc = locate(A.x + x_distance, A.y + y_distance, TO.z) // calculate the new place
+				if(!A.Move(newloc) && newloc) // if the atom, for some reason, can't move, FORCE them to move! :) We try Move() first to invoke any movement-related checks the atom needs to perform after moving
+					A.forceMove(newloc)
+
+				spawn()
```
Comment (farie82) [inline code/game/objects/effects/anomalies.dm:226]: Best to use a timer here

File: code/game/objects/effects/anomalies.dm [hunk 9]
```diff
@@ -60,72 +91,188 @@
...
+							var/obj/blueeffect = new /obj(src)
+							blueeffect.screen_loc = "WEST,SOUTH to EAST,NORTH"
+							blueeffect.icon = 'icons/effects/effects.dmi'
+							blueeffect.icon_state = "shieldsparkles"
+							blueeffect.layer = FLASH_LAYER
+							blueeffect.plane = FULLSCREEN_PLANE
+							blueeffect.mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+							M.client.screen += blueeffect
+							sleep(20)
+							M.client.screen -= blueeffect
```
Comment (farie82) [inline code/game/objects/effects/anomalies.dm:239]: M or M.client can be null at this point.

File: code/modules/events/anomaly.dm [hunk 10]
```diff
@@ -1,33 +1,19 @@
...
 	if(!impact_area)
-		setup(safety_loop)
+		CRASH("No valid areas for anomaly found.")
 	var/list/turf_test = get_area_turfs(impact_area)
 	if(!turf_test.len)
-		setup(safety_loop)
+		CRASH("Anomaly : No valid turfs found for [impact_area] - [impact_area.type]")
 
-/datum/event/anomaly/announce()
+/datum/event/anomaly/announce(fake)
```
Comment (farie82) [inline code/modules/events/anomaly.dm:13]: fake is not used nor a parameter in the base proc

**Pull request closed without merging**